File: app/src/main/java/com/mutant/godutch/Fragment/GroupsFragment.java
package com.mutant.godutch.Fragment;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ListView;

import com.firebase.client.ChildEventListener;
import com.firebase.client.DataSnapshot;
import com.firebase.client.Firebase;
import com.firebase.client.FirebaseError;
import com.mutant.godutch.R;

/**
 * Created by evanfang102 on 2017/3/30.
 */

public class GroupsFragment extends Fragment {

    ListView mListViewGroup;
    ArrayAdapter<String> mAdapter;

    public GroupsFragment() {
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_groups, container, false);
        setupGroups(view);
        return view;
    }

    private void setupGroups(View view) {
        mListViewGroup = (ListView) view.findViewById(R.id.listView_group);
        mAdapter = new ArrayAdapter<>(getContext(), android.R.layout.simple_list_item_1, android.R.id.text1);
        mListViewGroup.setAdapter(mAdapter);
        setupFirebase();
    }

    public static final String FIREBASE_URL = "https://godutch-c22a5.firebaseio.com/group";

    private void setupFirebase() {
        Firebase.setAndroidContext(getContext());
        new Firebase(FIREBASE_URL).addChildEventListener(new ChildEventListener() {
            @Override
            public void onChildAdded(DataSnapshot dataSnapshot, String s) {
                mAdapter.add((String) dataSnapshot.child("name").getValue());
                mAdapter.notifyDataSetChanged();
            }

            @Override
            public void onChildChanged(DataSnapshot dataSnapshot, String s) {

            }

            @Override
            public void onChildRemoved(DataSnapshot dataSnapshot) {
                mAdapter.remove((String) dataSnapshot.child("name").getValue());
                mAdapter.notifyDataSetChanged();
            }

            @Override
            public void onChildMoved(DataSnapshot dataSnapshot, String s) {

            }

            @Override
            public void onCancelled(FirebaseError firebaseError) {

            }
        });
    }
}

File: app/src/main/java/com/mutant/godutch/fragment/GroupsFragment.kt
package com.mutant.godutch.fragment

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.*
import com.mutant.godutch.AdapterGroup
import com.mutant.godutch.NewGroupActivity
import com.mutant.godutch.R
import com.mutant.godutch.model.Group
import kotlinx.android.synthetic.main.fragment_groups.*
import java.util.*

/**
 * Created by evanfang102 on 2017/3/30.
 */

class GroupsFragment : Fragment() {

    var mAdapterGroup: AdapterGroup? = null
    var firebaseUser = FirebaseAuth.getInstance().currentUser

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater!!.inflate(R.layout.fragment_groups, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupGroups()
        setupFirebase()
        setupFabNewGroup(view)
    }

    private fun setupFabNewGroup(view: View) {
        view.findViewById(R.id.fab_new_group).setOnClickListener { startActivity(NewGroupActivity.getIntent(activity)) }
    }

    private fun setupGroups() {
        mAdapterGroup = AdapterGroup(activity, ArrayList<Group>())
        val MyLayoutManager = LinearLayoutManager(activity)
        MyLayoutManager.orientation = LinearLayoutManager.VERTICAL
        recycler_view_groups.adapter = mAdapterGroup
        recycler_view_groups.layoutManager = MyLayoutManager
    }

    private fun setupFirebase() {
        val databaseGroupsMapping: DatabaseReference? = FirebaseDatabase.getInstance().reference.
                child("groups_mapping").child(firebaseUser?.uid)
        databaseGroupsMapping?.addChildEventListener(object : ChildEventListener {
            override fun onChildMoved(p0: DataSnapshot?, p1: String?) {
            }

            override fun onChildChanged(dataSnapshot: DataSnapshot?, p1: String?) {
                // TODO
            }

            override fun onChildAdded(dataSnapshot: DataSnapshot?, p1: String?) {
                fetchGroupsData((dataSnapshot?.value as String))
            }

            override fun onChildRemoved(p0: DataSnapshot?) {
            }

            override fun onCancelled(dataSnapshot: DatabaseError?) {
            }

        })
    }

    private fun fetchGroupsData(groupsKey: String) {
        val databaseGroups: DatabaseReference? = FirebaseDatabase.getInstance().reference.
                child("groups").child(groupsKey)
        databaseGroups?.orderByKey()?.addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                var group = dataSnapshot.getValue(Group::class.java)
                group.key = groupsKey
                mAdapterGroup?.addItem(group)
                recycler_view_groups.scrollToPosition(0)
            }

            override fun onCancelled(p0: DatabaseError?) {
            }
        })
    }
}
